src/inmanta/server/agentmanager.py

```diff
@@ -187,42 +188,77 @@ def get_agent_client(self, tid: uuid.UUID, endpoint: str) -> Optional[ReturnClie
             if agent is not None:
                 return agent
             else:
-                agent = await self._create_default_agent(env, nodename)
-                return agent
+                return await self._create_default_agent(env, nodename)
 
     async def _create_default_agent(self, env: data.Environment, nodename: str) -> data.Agent:
         saved = data.Agent(environment=env.id, name=nodename, paused=False)
         await saved.insert()
-        await self._verify_reschedule(env, [nodename])
+
+        key = (env.id, nodename)
+        session = self.tid_endpoint_to_session.get(key, None)
+        if session is not None:
+            await self._set_primary(env, [(nodename, session)], datetime.now())
+
         return saved
 
     async def _register_session(self, session: protocol.Session, now: datetime) -> None:
         async with self.session_lock:
             tid = session.tid
             sid = session.get_id()
-            nodename = session.nodename
+
+            agent_statuses = await data.Agent.get_statuses(tid, session.endpoint_names)
 
             self.sessions[sid] = session
 
-            env = await data.Environment.get_by_id(tid)
-            if env is None:
-                LOGGER.warning("The environment id %s, for agent %s does not exist!", tid, sid)
+            endpoints_with_new_primary: List[Tuple[str, Optional[protocol.Session]]] = []
+            for endpoint in session.endpoint_names:
+                if (tid, endpoint) not in self.tid_endpoint_to_session and agent_statuses[endpoint] != "paused":
+                    LOGGER.debug("set session %s as primary for agent %s in env %s", sid, endpoint, tid)
+                    self.tid_endpoint_to_session[(tid, endpoint)] = session
+                    self.add_background_task(session.get_client().set_state(endpoint, True))
+                    endpoints_with_new_primary.append((endpoint, session))
+
+            await self._log_session_creation_to_db(tid, endpoints_with_new_primary, session, now)
```

Comment:
Should this be a background task? Now it is under the session lock, blocking all other sessions.

Reply:
Indeed. In the previous implementation, database updates happened under session lock. Now this can happen in the background.